File: agents/triage/src/services/llm_service.py
import httpx
import json
from src.config import settings
from src.utils.logging import setup_logger

logger = setup_logger(__name__)

class LLMService:
    def __init__(self, model: str = None):
        self.base_url = settings.INFERENCE_SERVER_URL
        self.model = model or settings.LLM_MODEL
        logger.info(f"LLM Service Initialized with Model: {self.model} | Base URL: {self.base_url}")

    async def generate_response(self, prompt: str, system_prompt: str = None, temperature: float = 0.0) -> str:
        """
        Legacy method for single prompt generation.
        """
        return await self.chat_completion([
            {"role": "system", "content": system_prompt or ""},
            {"role": "user", "content": prompt}
        ], options={"temperature": temperature})

    async def chat_completion(self, messages: list, options: dict = None) -> str:
        """
        Calls Ollama Chat API.
        Endpoint: POST /api/chat
        """
        url = f"{self.base_url}/api/chat"
        
        valid_messages = [m for m in messages if m.get("content")]

        default_options = {
            "temperature": 0.3, 
        }
        if options:
            default_options.update(options)

        payload = {
            "model": self.model,
            "messages": valid_messages,
            "stream": False,
            "options": default_options
        }

        try:
            async with httpx.AsyncClient() as client:
                logger.debug(f"Calling LLM Chat with {len(messages)} messages")
                response = await client.post(url, json=payload, timeout=60.0) # Increased timeout
                response.raise_for_status()
                data = response.json()
                return data["message"]["content"]
        except httpx.HTTPError as e:
            logger.error(f"LLM Chat generation failed: {e}")
            raise
    async def extract_entities(self, text: str, schema: dict) -> dict:
        """
        Extracts structured entities from text based on a schema.
        Returns a dictionary with the extracted data.
        """
        system_prompt = f"""
        You are a precise Data Extraction Engine.
        Your goal is to extract entities from the user's text based on the keys below.
        Return ONLY valid JSON.
        
        Keys to extract:
        {json.dumps(schema, indent=2)}
        
        Rules:
        - If a value is missing, use null.
        - Do not assume or hallucinate values.
        - For 'action', use normalized verbs (liberar, desbloquear, criar, remover).
        
        Example Input: "Liberar acesso ao SOE para o João"
        Example Output: {{"system": "SOE", "action": "liberar", "beneficiary_name": "João", "beneficiary_id": null}}
        """
        
        try:
            response_text = await self.generate_response(
                prompt=text,
                system_prompt=system_prompt,
                temperature=0.0
            )
            
            # Clean response to ensure valid JSON (remove markdown code blocks if any)
            clean_text = response_text.replace("```json", "").replace("```", "").strip()
            
            return json.loads(clean_text)
            
        except json.JSONDecodeError as e:
            logger.error(f"Failed to decode JSON from LLM: {e}")
            logger.debug(f"Raw response: {response_text}")
            return {}
        except Exception as e:
            logger.error(f"Entity extraction failed: {e}")
            return {}

File: glpi-ticket-agent/src/services/llm_service.ts
import fetch from 'node-fetch';
import { configManager } from '../config';
import { jsonrepair } from 'jsonrepair';

export class LLMService {
    private config = configManager.getLLMConfig();

    /**
     * Completes a prompt using the configured LLM.
     * Returns the raw string response.
     */
    async complete(prompt: string, systemPrompt?: string): Promise<string> {
        // Normalize URL: Remove trailing /v1 or / if present to avoid duplication
        const baseUrl = this.config.baseUrl.replace(/\/v1\/?$/, '').replace(/\/$/, '');
        const url = `${baseUrl}/v1/chat/completions`;

        // Fallback system prompt
        const sys = systemPrompt || "You are a helpful assistant.";

        const body = {
            model: this.config.model,
            messages: [
                { role: "system", content: sys },
                { role: "user", content: prompt }
            ],
            temperature: this.config.temperature,
            max_tokens: 1024,
            stream: false
        };

        try {
            console.log(`[LLMService] Calling ${url} with model ${this.config.model}`);
            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.config.apiKey}`
                },
                body: JSON.stringify(body)
            });

            if (!response.ok) {
                const errText = await response.text();
                throw new Error(`LLM API Error ${response.status}: ${errText}`);
            }

            const data = await response.json() as any;

            // OpenAI format
            const content = data.choices?.[0]?.message?.content || "";
            return content.trim();

        } catch (error) {
            console.error("[LLMService] Failed to complete:", error);
            return "";
        }
    }

    /**
     * Completes a prompt and parses the result as JSON.
     * Uses jsonrepair to fix common LLM syntax errors.
     */
    async completeJson<T>(prompt: string, schemaDescription?: string): Promise<T | null> {
        const systemPrompt = `You are a JSON-only API. You must return a valid JSON object matching the requested schema. Do not include markdown blocks like \`\`\`json. Just the raw JSON.
    
    ${schemaDescription ? "Expected Schema:\n" + schemaDescription : ""}
    `;

        const raw = await this.complete(prompt, systemPrompt);
        if (!raw) return null;

        try {
            // Clean up markdown code blocks if present (common Llama artifact)
            let clean = raw.replace(/```json/g, '').replace(/```/g, '').trim();

            // Use jsonrepair to handle missing quotes, trailing commas, etc.
            const parsed = JSON.parse(jsonrepair(clean));
            return parsed as T;
        } catch (e) {
            console.error("[LLMService] JSON Parse Error on:", raw, e);
            return null;
        }
    }
}

export const llmService = new LLMService();
